Python/specialize.c

```diff
@@ -2560,14 +2560,35 @@ static int
 binary_op_extended_specialization(PyObject *lhs, PyObject *rhs, int oparg,
                                   _PyBinaryOpSpecializationDescr **descr)
 {
-    size_t n = sizeof(binaryop_extend_descrs)/sizeof(_PyBinaryOpSpecializationDescr);
-    for (size_t i = 0; i < n; i++) {
-        _PyBinaryOpSpecializationDescr *d = &binaryop_extend_descrs[i];
+    typedef _PyBinaryOpSpecializationDescr descr_type;
+    size_t size = sizeof(binaryop_extend_builtins)/sizeof(descr_type);
+    for (size_t i = 0; i < size; i++) {
+        descr_type *d = (descr_type *)&binaryop_extend_builtins[i];
+        assert(d != NULL);
+        assert(d->guard != NULL);
         if (d->oparg == oparg && d->guard(lhs, rhs)) {
             *descr = d;
             return 1;
         }
     }
+
+    if (Py_TYPE(lhs)->tp_binop_specialize != NULL) {
+        int ret = Py_TYPE(lhs)->tp_binop_specialize(lhs, rhs, oparg, descr);
+        if (ret < 0) {
+            return -1;
+        }
+        if (ret == 1) {
+            if (*descr == NULL) {
+                PyErr_Format(
```

Comment:
Maybe chain exceptions if possible here? it could be helpful for extensions to know that they set an exception but returned 1.

Reply:
It's a bit awkward to do, and I don't think it's that useful. If the guard/action functions are buggy then the context in which specialisation was kicked off is unlikely to be relevant.